connectors/sync_job_runner.py

```diff
@@ -0,0 +1,151 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+import asyncio
+import time
+
+from connectors.byoc import JobStatus
+from connectors.es import Mappings
+from connectors.logger import logger
+
+
+class SyncJobRunningError(Exception):
+    pass
+
+
+class SyncJobRunner:
+    """The class to run a sync job."""
+
+    def __init__(
+        self,
+        source_klass,
+        sync_job,
+        connector,
+        elastic_server,
+        bulk_options,
+    ):
+        self.source_klass = source_klass
+        self.sync_job = sync_job
+        self.job_id = self.sync_job.id
+        self.connector = connector
+        self.connector_id = self.connector.id
+        self.elastic_server = elastic_server
+        self.bulk_options = bulk_options
+        self._start_time = None
+        self.running = False
+
+    async def execute(self):
+        if self.running:
+            raise SyncJobRunningError(f"Sync job #{self.job_id} is already running.")
+
+        self.running = True
+        if not await self._claim_job():
+            logger.error(f"Unable to claim job #{self.job_id}")
+            return
+
+        try:
+            data_provider = self.source_klass(self.sync_job.configuration)
+            if not await data_provider.changed():
```

Comment:
Do we actually implement this method in a meaningful way somewhere? It's always `true` in the base class and `directory.py` implements it also always returning `true`.

Reply:
the initial intent was to bypass a sync if we know nothing has changed. I used it in the first version of the Mongo connector. In S3/GCP it can be implemented by looking a the bucket last changed timestamp